src/controller/python/chip/clusters/Attribute.py

```diff
@@ -723,9 +724,12 @@ def handleEventData(self, header: EventHeader, path: EventPath, data: bytes, sta
                 Header=header, Data=eventValue, Status=chip.interaction_model.Status(status))
             self._events.append(eventResult)
 
-            if (self._subscription_handler is not None):
-                self._subscription_handler.OnEventChangeCb(
-                    eventResult, self._subscription_handler)
+            if self._subscription_handler is not None:
+                callback = self._subscription_handler.OnEventChangeCb
+                if inspect.iscoroutinefunction(callback):
```

Comment:
This call is quite expensive. Scheduling it as a task just to forward a piece of information can be quite some overhead. Would it be possible to allow decorating functions that they need to be run inside the event loop ? In Home Assistant we have a [`@callback`](https://github.com/home-assistant/core/blob/dev/homeassistant/core.py#L220-L228) decorator to indicate that a function is to be called in the event loop.

Reply:
> This call is quite expensive. Scheduling it as a task just to forward a piece of information can be quite some overhead. Would it be possible to allow decorating functions that they need to be run inside the event loop ? In Home Assistant we have a [@callback](https://github.com/home-assistant/core/blob/dev/homeassistant/core.py#L220-L228) decorator to indicate that a function is to be called in the event loop.

Looking into the implementation of `iscoroutinefunction` it seems quite simple tough. 

Quick test:
```python
import timeit
setup="""
from inspect import iscoroutinefunction
async def test():
    pass
"""
timeit.timeit("iscoroutinefunction(test)",setup=setup)

setup="""\
def callback(func):
    setattr(func, "_hass_callback", True)
    return func
def is_callback(func) -> bool:
    return getattr(func, "_hass_callback", False) is True
@callback
async def test():
    pass
"""
timeit.timeit("is_callback(test)",setup=setup)
```

`iscoroutinefunction` is ~26s vs. `is_callback` ~5.4s, so quite a bit faster indeed.

There are actually two question here:
- Should we callback on the event loop or in the Thread the SDK calls us?
- Should we support coroutine functions (as in `async def...`)

In the end, both handled by the callback itself, essentially leave it up to the user of the read/write attribute APIs.

I think my main gripe here is that different callbacks have different semantics right now. And they are also not documented :cry: 

I probably gets a bit philosophical. What do you think how this API should behave? :thinking: